crates/core/src/replication/connection.rs

```diff
@@ -0,0 +1,124 @@
+// TODO(lucio): Move this to `remote/mod.rs`
+
+use std::sync::Arc;
+
+use parking_lot::Mutex;
+
+use crate::replication::pb::{execute_results::State as RemoteState, query_result::RowResult};
+use crate::{
+    params::Params, replication::Writer, Error, Result, Statement, Transaction, TransactionBehavior,
+};
+
+use crate::v2::{Conn, LibsqlConnection};
+
+use super::parser;
+
+pub struct RemoteConnection {
+    local: LibsqlConnection,
+    writer: Writer,
+    state: Arc<Mutex<State>>,
+}
+
+#[derive(Default)]
+struct State {
+    remote_state: RemoteState,
+    is_autocommit: bool,
+    changes: u64,
+    last_insert_rowid: i64,
+}
+
+impl RemoteConnection {
+    pub(crate) fn new(local: LibsqlConnection, writer: Writer) -> Self {
+        let state = Arc::new(Mutex::new(State::default()));
+        Self {
+            local,
+            writer,
+            state,
+        }
+    }
+
+    fn is_state_init(&self) -> bool {
+        matches!(dbg!(self.state.lock().remote_state), RemoteState::Init)
+    }
+}
+
+#[async_trait::async_trait]
+impl Conn for RemoteConnection {
+    async fn execute(&self, sql: &str, params: Params) -> Result<u64> {
+        let stmts = parser::Statement::parse(sql).collect::<Result<Vec<_>>>()?;
+
+        let is_read_only = stmts
+            .iter()
+            .fold(false, |ro, s| if s.is_read_only() { true } else { ro });
+
+        if self.is_state_init() && is_read_only {
+            // TODO(lucio): See if we can arc the params here to cheaply clone
+            // or convert the inner bytes type to an Arc<[u8]>
+            let changes = self.local.execute(dbg!(sql), params.clone()).await;
+
+            // Check if we are left in a TXN state
+            if !dbg!(self.local.is_autocommit()) {
+                self.local.execute("ROLLBACK", Params::None).await?;
```

Comment:
Is the intention of this block to allow executing reads locally on startup? I don't follow why this ROLLBACK is here. If somebody calls execute 3 times like that:
```
execute("BEGIN")
execute("SELECT v FROM t")
execute("SELECT v2 FROM t2")
execute("ROLLBACK")
```
, we'll end up implicitly calling `ROLLBACK` after the first `BEGIN`, and then the following selects would happen out of a single read transaction like the user intended, right? Is it specifically a workaround for what javscript is doing with their `SELECT 1` during startup?

Reply:
Also, where does the state stop being Init? Is this path expected to be taken only on startup or generally during reads?

Reply:
So this is to work around interactive txns. So the idea is that if you want to do a read only txn you want to batch every call into one execute call where it ends the transaction within the same execute call. If `BEGIN` is sent it will trigger the case below where the connection is still in not in autocommit and it will forward the BEGIN to the primary. This then will trigger the TXN state that we will then send all following sql statements to the primary until we return to the init state by sending a `ROLLBACK` or `COMMIT`. This is the way that sqld does it. I think we can be smarter in the future here but this seems good enough now to support what we want.

The state will go from `Init` ->`Txn` when we send a `BEGIN` to the remote and will return back to `Init` when the remote sees a `COMMIT` or `ROLLBACK`.